src/field.cpp

```diff
@@ -367,6 +369,416 @@ Option<bool> filter::parse_filter_query(const string& simple_filter_query,
 
     return Option<bool>(true);
 }
+bool isOperator(const std::string& token) {
+    return token == "&&" || token == "||";
+}
+
+// https://en.wikipedia.org/wiki/Shunting_yard_algorithm
+Option<bool> toPostfix(std::queue<std::string>& tokens, std::queue<std::string>& postfix) {
+    std::stack<std::string> operatorStack;
+
+    while(!tokens.empty()) {
+        auto token = tokens.front();
+        tokens.pop();
+
+        if (isOperator(token)) {
+            // We only have two operators &&, || having the same precedence and both being left associative.
+            while (!operatorStack.empty() && operatorStack.top() != "(") {
+                postfix.push(operatorStack.top());
+                operatorStack.pop();
+            }
+
+            operatorStack.push(token);
+        } else if (token == "(") {
+            operatorStack.push(token);
+        } else if (token == ")") {
+            while (!operatorStack.empty() && operatorStack.top() != "(") {
+                postfix.push(operatorStack.top());
+                operatorStack.pop();
+            }
+
+            if (operatorStack.empty() || operatorStack.top() != "(") {
+                return Option<bool>(400, "Could not parse the filter query.");
```

Comment:
We could probably given a more helpful message like `Could not parse the filter query: unbalanced parentheses.`

In other places also (e.g. in `toParseTree`), if we can give an explicit reasoning for parse error we should do so.